q2cli/handlers.py

```diff
@@ -0,0 +1,214 @@
+
+class ResultHandler(GeneratedHandler):
+    prefix = 'o_'
+
+    def __init__(self, extension, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.extension = extension
+
+    def get_click_options(self):
+        yield click.Option(['--' + self.normalized_name],
+                           type=click.Path(exists=False, dir_okay=False),
+                           help="Output %r" % self.semtype)
+
+    def get_value(self, arguments, fallback=None):
+        path = self._locate_value(arguments, fallback)
+        if not path.endswith(self.extension):
+            path += self.extension
+
+        return path
+
+
+def parameter_handler_factory(name, semtype, default=NoDefault):
+    ast = semtype.to_ast()
+    if ast['name'] == 'Metadata':
+        return MetadataHandler(name)
+    elif ast['name'] == 'MetadataCategory':
+        return MetadataCategoryHandler(name)
+    return RegularParameterHandler(name, semtype, default=default)
+
+
+class MetadataHandler(Handler):
+    def __init__(self, name):
+        super().__init__(name, prefix='m_')
+        self.click_name += '_file'
+
+    def get_click_options(self):
+        yield click.Option(['--' + self.normalized_name],
+                           type=click.Path(exists=True, dir_okay=False),
+                           help='Metadata mapping file')
+
+    def get_value(self, arguments, fallback=None):
+        value = self._locate_value(arguments, fallback)
+        return qiime.Metadata.load(value)
+
+
+class MetadataCategoryHandler(Handler):
+    def __init__(self, name):
+        self.name = name
+        self.click_names = ['m_%s_file' % name, 'm_%s_category' % name]
+        self.normalized_names = [
+            q2cli.util.normalize_name(n) for n in self.click_names]
+        self.normalize_name = self.normalized_names[0]
+
+    def get_click_options(self):
+        yield click.Option(['--' + self.normalized_names[0]],
+                           type=click.Path(exists=True, dir_okay=False),
+                           help='Metadata mapping file')
+
+        yield click.Option(['--' + self.normalized_names[1]],
+                           type=str,
+                           help='Category from metadata mapping file')
+
+    def get_value(self, arguments, fallback=None):
+        values = []
+        for click_name, normalized_name in zip(self.click_names,
+                                               self.normalized_names):
+            value = self._locate_value(
+                arguments, fallback, click_name=click_name,
+                normalized_name=normalized_name)
+            values.append(value)
+
+        return qiime.MetadataCategory.load(*values)
+
+
+class RegularParameterHandler(GeneratedHandler):
+    prefix = 'p_'
+
+    def get_type(self):
+        mapping = {
+            'Int': int,
+            'Str': str,
+            'Float': float,
+            'Color': str
+        }
+        predicate = self.ast['predicate']
+        if predicate:
```

Comment:
Comment that we are only handling this one case and expect this to be changed significantly in the future.